### ISSUE TITLE
use alternative optimizer

### ISSUE THREAD

**Issue by Borda** [ultralytics/yolov5#929]
## 🚀 Feature

allow using alternative optimizer, especially in fine-tuning/evolve mode

## Motivation
cor some cases the default optimizer may not be optimal and having option to define own would be benefitial

## Alternatives

https://pypi.org/project/torch-optimizer - this is quite nice package with large optimizes zoo with the same API as pyTorch

## Additional context

the optimizer can be defined as string/dict in config
```yaml
optimizer: Adam
```
or
```yaml
optimizer:
    name: Adam
    lr: 0.0005
```

**Comment by Borda** [comment]
any progress here?

**Comment by glenn-jocher** [comment]
@Borda this is an interesting topic. In my experiments I've seen Adam can work well on smaller custom datasets, and can provide good initial results on larger datasets, whereas SGD tends to outperform in the long run, especially on larger datasets, and seems to generalize better to real world results.

PyTorch makes it easy to swap out different optimizers, but in practice the learning rate and other hyps are severely optimizer specific, so real world optimizer changes require a lot of experimentation and experience. The main strategy I have when testing Adam is to start at a 0.1x LR compared to what I used for SGD, which is shown in the hyps comment here:
https://github.com/ultralytics/yolov5/blob/0ada058f6359b9c76569e9b501a5da7da0a11d74/data/hyp.scratch.yaml#L6

This is a gross oversimplification of the challenges though, I'm not really sure what to recommend here. Making a switch very easy will result in many users trying one or another (with zero hyp changes), and then assuming the new results to be the final say on whether one optimizer works better than another. That being said though, if Adam works better on most use cases (new users with small custom datasets), perhaps we may want to even use it as the new default, and leave the experts to switch to SGD if they want. What do you think?

**Comment by glenn-jocher** [comment]
@Borda this got me curious, I'm going to test Adam on VOC and a custom dataset. I think there may be significant overlap between the typical scenario Adam excels at and the types of datasets users typically arrive here with.

Unfortunately I see many users underestimate the size and variety of data required for best generalization in their domains, which leads to understandably poor training results sometimes, which can discourage new users. May adam could play a part in helping there, though it's still an open question/concern whether val mAP improvements Adam provides correlate well to real world improvements, as my own experiences and what I've read in the literature seem to indicate otherwise.
